enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,443 @@
+example, if the hardware MTU is 1500, then the auto-probed cluster MTU would be
+1400 for OVN Kubernetes and 1450 for OpenShift SDN. During the migration, this
+MTU mismatch would break some cross-CNI traffic.
+
+To avoid this issue, before starting the migration, users need to decrease the
+cluster MTU by 50 bytes following this instruction:
+[changing-cluster-network-mtu](https://docs.openshift.com/container-platform/4.13/networking/changing-cluster-network-mtu.html).
+
+### The Traffic Path
+
+#### Packets going from OpenShift SDN to OVN Kubernetes
+
+On the SDN side, it doesn't need to know if the peer node is a SDN node or an
+OVN node. We reuse the existing VXLAN tunnel rules on the SDN side.
+- Egress NetworkPolicy rules and service proxy happen as normal.
+- When the packet reaches table 90, it will hit a "send via vxlan" rule that was
+  generated based on a HostSubnet object.
+
+On the OVN side:
+- OVN accepts the packet via the VXLAN tunnel, ignores the VNID set by SDN, and
+  then just routes it normally.
+- Ingress NetworkPolicy processing will happen when the packet reaches the
+  destination pod's switch port, just like normal.
+  - Our NetworkPolicy rules are all based on IP addresses, not "logical input
+    port", etc., so it doesn't matter that the packets came from outside OVN and
+    have no useful OVN metadata.
+
+#### Packets going from OVN Kubernetes to OpenShift SDN
+
+On the OVN side:
+- The packet just follows the same path as the hybrid overlay.
+
+On the SDN side:
+- We have to change ingress NetworkPolicy processing to be based entirely on pod
+  IPs rather than using namespace VNIDs since packets from OVN nodes won't have
+  the VNID set. There is already code to generate the rules that way, though,
+  because egress NetworkPolicy already works that way.
+
+### Workflow Description
+
+1. The admin kicks off the migration process by updating the custom resource
+   `network.config`:
+   
+   ```bash
+   $ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/live-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'
+   ```
+   
+   CNO will set the `LiveMigrationProgressing` condition to TRUE in the `status` of
+   the network.config CR. If there are any unsupported features (refer to the
+   [Limitations](#limitations) section) enabled in the cluster. CNO will
+   set an the `LiveMigrationProgressing` condition to `False` with warning message
+   and won't start the live migration. User shall disable these unsupported
+   feature before starting the live migration.
+
+2. CNO will check the current hardware MTU and the cluster MTU to see if there
+   is enough free space for OVN-K overlay overhead. If not, sets the
+   `.status.migration.mtu` of network.config CR with MTU of OVN-k, e.g.:
+   - `.status.migration.mtu.network.from` 1450 `to` 1400
+   - `.status.migration.mtu.machine.from` 1500 `to` 1500
```

Comment:
Is this dynamic based on what the current settings are? Will it be restored afterwards to the previous setting? While most clusters within a cloud will have similar settings, the use of certain features (e.g. AWS Local Zones) necessitates a lower network overlay MTU (ref https://docs.openshift.com/container-platform/4.13/installing/installing_aws/installing-aws-localzone.html#edge-machine-pools-aws-local-zones_installing-aws-localzone)

Reply:
Yes. The new cluster MTU will be calculated based on the value of the machine MTU and the current cluster MTU.

Reply:
For AWS localzone instance, the underlay/machine MTU is 1300. It will not be changed by the migration. The migration only changes the overlay/cluster MTU which is the MTU used by the pod interfaces. So in your case, the calculated cluster MTU is 1250 for O-SDN, and 1200 for OVN-K. So after the migration. The cluster will be 1200.

Reply:
Okay, I understand now a bit more clearly although not all of your statement about local zones is correct. The machines themselves have an MTU of 9001 like other AWS nodes, but you do need to lower the overlay network MTU because the inter-zone networking in AWS for local zones to non-local zones is 1300, minus any overhead.

So for an SDN cluster with local zone nodes, I would expect the cluster/overlay MTU to be 1250, and for OVN it would be 1200 like you indicate. And for a default installed AWS cluster with no local zones, the overlay network MTU would be 8951 for SDN, and 8001 for OVN.

Thanks for clarifying!